### ISSUE TITLE
[TPU] Intermittent NaN in tied embedding gradient

### ISSUE THREAD

**Issue by martin-marek** [jax-ml/jax#35409]
## Description

A language model uses a single **tied embedding matrix** for two purposes: (1) looking up input token embeddings (`gather`), and (2) projecting hidden states to logits (`dot_general`). During backprop, this means the embedding receives gradients from two different paths, let's call them "sparse" and "dense". When both paths contribute to the same parameter inside a single training step on TPU, the embedding intermittently contains NaN/inf values.

## Reproduction

The script below implements a small transformer with a tied embedding and runs a single training step (forward + backward + AdamW update) repeatedly with **identical frozen inputs** &mdash; same weights, same optimizer state, same batches every call. It checks outputs for finiteness after each call.

Four modes isolate what triggers the failure:

| Command | What changes | Result |
|------|-------------|--------|
| `python repro.py --mode full --remat` | Tied embedding, both grad paths active | **Intermittent NaN** |
| `python repro.py --mode sparse_only --remat` | Tied embedding, dense grad blocked with `stop_gradient` | Stable |
| `python repro.py --mode dense_only --remat` | Tied embedding, sparse grad blocked with `stop_gradient` | Stable |
| `python repro.py --mode full` | No gradient checkpointing | Stable |

To reproduce these examples, save the code below as `repro.py`:

<details>
<summary><b>repro.py</b> (click to expand)</summary>

```python
import argparse
import json
from functools import partial

import jax
import jax.numpy as jnp
import optax
from jax.sharding import AxisType, PartitionSpec as P

USED_LAYERS, TOTAL_LAYERS = 6, 16 # <- having unused layers seems necessary to trigger the bug
D_MODEL, FFN, HEADS, KV_HEADS, VOCAB = 2048, 8192, 16, 16, 100352
B1, T1, B2, T2 = 512, 32, 16, 1024 # <- using different shapes for each batch seems necessary to trigger the bug
LR = 1e-6


def rms(x, g, eps=1e-5, axis=-1):
    return (
        g * x
        / jnp.sqrt(jnp.mean(x.astype(jnp.float32) ** 2, axis=axis, keepdims=True) + eps)
    ).astype(x.dtype)


def tree_finite(tree):
    leaves = jax.tree.leaves(tree)
    return all((not hasattr(x, "dtype")) or bool(jnp.isfinite(x).all()) for x in leaves)


def make_batch(b, t, seed):
    k = jax.random.key(seed)
    tok = jax.random.randint(k, (b, t), 0, VOCAB, dtype=jnp.int32)
    ln = jax.random.randint(jax.random.fold_in(k, 1), (b,), max(2, t // 3), t + 1)
    return tok


def make_weights(seed):
    rng = jax.random.key(seed)
    hd = D_MODEL // HEADS

    mats = {
        "wq": ((HEADS, hd, D_MODEL), P("model", "data")),
        "wo": ((D_MODEL, HEADS, hd), P("data", "model")),
        "wg": ((FFN, D_MODEL), P("model", "data")),
        "wu": ((FFN, D_MODEL), P("model", "data")),
        "wd": ((D_MODEL, FFN), P("data", "model")),
        "wk": ((KV_HEADS, hd, D_MODEL), P("model", "data")),
        "wv": ((KV_HEADS, hd, D_MODEL), P("model", "data")),
    }
    ones = {"qg": (HEADS, hd), "an": (D_MODEL,), "fn": (D_MODEL,), "kg": (KV_HEADS, hd)}

    num_keys = 1 + TOTAL_LAYERS * len(mats)
    keys = iter(jax.random.split(rng, num_keys))

    w = {
        "embed": jax.device_put(jax.random.normal(next(keys), (VOCAB, D_MODEL), dtype=jnp.float32), P("model", "data")),
        "final_norm": jax.device_put(jnp.ones((D_MODEL,), jnp.float32), P()),
    }
    
    for i in range(TOTAL_LAYERS):
        p = f"l{i}."
        for n, (shape, spec) in mats.items():
            w[p + n] = jax.device_put(jax.random.normal(next(keys), shape, dtype=jnp.float32), spec)
        for n, shape in ones.items():
            w[p + n] = jax.device_put(jnp.ones(shape, jnp.float32), P())
            
    return w


def make_forward(mode, remat):
    @jax.jit # <- jit here seems necessary to trigger the bug
    def forward(tokens, w):
        emb = w["embed"]
        emb_in = emb if mode != "dense_only" else jax.lax.stop_gradient(emb)
        x = emb_in.at[tokens, :].get(out_sharding=P("data", None, None)).astype(jnp.bfloat16)
        for i in range(USED_LAYERS):
            p = f"l{i}."
            def layer(y):
                q = jnp.einsum("btd,nhd->btnh", y, w[p+"wq"], preferred_element_type=y.dtype, out_sharding=P("data", None, "model", None))
                q = rms(q, w[p+"qg"], axis=(-2, -1))
                o = jnp.einsum("btnh,dnh->btd", q, w[p+"wo"], preferred_element_type=y.dtype, out_sharding=P("data", None, None))
                y = y + rms(o, w[p+"an"])
                g = jax.nn.silu(jnp.einsum("btd,fd->btf", y, w[p+"wg"], preferred_element_type=jnp.float32, out_sharding=P("data", None, "model")))
                u = jnp.einsum("btd,fd->btf", y, w[p+"wu"], preferred_element_type=y.dtype, out_sharding=P("data", None, "model"))
                d = jnp.einsum("btf,df->btd", g*u, w[p+"wd"], preferred_element_type=y.dtype, out_sharding=P("data", None, None))
                return y + rms(d, w[p+"fn"])
            x = jax.remat(layer)(x) if remat else layer(x)
        x = rms(x, w["final_norm"])
        emb_out = emb if mode != "sparse_only" else jax.lax.stop_gradient(emb)
        return jnp.einsum("btd,vd->btv", x, emb_out, preferred_element_type=x.dtype, out_sharding=P("data", None, "model"))
    return forward


def loss_fn(fwd, w, a, b):
    def one(x):
        y = jnp.roll(x, -1, axis=1)
        lsm = jax.nn.log_softmax(fwd(x, w).astype(jnp.float32))
        ce = -jnp.sum(jax.nn.one_hot(y, lsm.shape[-1]) * lsm, axis=-1)
        return ce.mean()
    return one(a) + one(b) # <- using two batches seems necessary to trigger the bug


@partial(jax.jit, static_argnames=("fwd", "opt"))
def step(fwd, wt, opt, st, a, b):
    l, g = jax.value_and_grad(lambda ww: loss_fn(fwd, ww, a, b))(wt)
    u, st2 = opt.update(g, st, wt)
    return optax.apply_updates(wt, u), st2, l


def run_mode(mode, seeds, trials, remat):
    a, b = make_batch(B1, T1, 100), make_batch(B2, T2, 200)
    opt = optax.adamw(optax.constant_schedule(LR), 0.9, 0.997, weight_decay=0.01)
    fwd = make_forward(mode, remat=remat)
    results = []
    for seed in range(seeds):
        w = make_weights(seed)
        st = opt.init(w)
        bad_calls = 0
        for i in range(trials):
            w2, st2, l = step(fwd, w, opt, st, a, b)
            if not tree_finite(w2):
                bad_calls += 1
        results.append({"seed": seed, "bad_calls": bad_calls, "trials": trials})
    return results


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--mode", choices=["full", "sparse_only", "dense_only"], default="full")
    ap.add_argument("--remat", action="store_true")
    ap.add_argument("--seeds", type=int, default=4)
    ap.add_argument("--trials", type=int, default=20)
    args = ap.parse_args()

    mesh = jax.make_mesh((jax.device_count(), 1), ("data", "model"), axis_types=(AxisType.Explicit, AxisType.Explicit))
    jax.set_mesh(mesh)
    print("env", {"jax": jax.__version__, "optax": optax.__version__,
                  "devices": [str(d) for d in jax.devices()],
                  "mode": args.mode, "remat": args.remat})

    results = run_mode(args.mode, args.seeds, args.trials, args.remat)
    for row in results:
        print(json.dumps(row))


if __name__ == "__main__":
    main()
```

</details>

## Output

Here's the number of non-finite updates I observed across various configs:

| Mode | remat | seed 1 | seed 2 | seed 3 | seed 4 | Total |
| :--- | :---: | :---: | :---: | :---: | :---: | :---: |
| `full` | `True` | 8/20 | 9/20 | 8/20 | 9/20 | **34/80** |
| `sparse_only` | `True` | 0/20 | 0/20 | 0/20 | 0/20 | 0/80 |
| `dense_only` | `True` | 0/20 | 0/20 | 0/20 | 0/20 | 0/80 |
| `full` | `False` | 0/20 | 0/20 | 0/20 | 0/20 | 0/80 |
| `sparse_only` | `False` | 0/20 | 0/20 | 0/20 | 0/20 | 0/80 |
| `dense_only` | `False` | 0/20 | 0/20 | 0/20 | 0/20 | 0/80 |

Note: numbers in the first row vary across runs, potentially even across machines.

## Model architecture

I intially discovered NaNs when training a model with Olmo2-1B architecture. However, for the purpose of this bug report, I wanted to find a simpler architecture that triggers the same bug. Strangely, I was unable to trigger the bug if I significantly simplified the architecture or made the model significantly smaller.

There seems to be lots of tiny details that if changed, the bug goes away. I've commented on a few in the included source code, examples include: remat, placement of jit, precision, and using two heterogeneous batches of data for loss computation.

## HLO context

The snippets below are provided in case they're useful for tracking down the issue. They're from the compiled HLO of the `full` mode executable.

<details>
<summary><b>StableHLO: how the embedding gradient is computed</b></summary>

The sparse gradient is a `scatter-add` into a zero buffer (the transpose of the input `gather`). It's then added to the dense gradient from the output projection:

```
    %cst_202 = stablehlo.constant dense<0.000000e+00> : tensor<f32>
    %2089 = stablehlo.broadcast_in_dim %cst_202, dims = [] : (tensor<f32>) -> tensor<100352x2048xf32>
    %2090 = "stablehlo.scatter"(%2089, %arg0, %2088) <{
        indices_are_sorted = false,
        scatter_dimension_numbers = #stablehlo.scatter<
            update_window_dims = [2], inserted_window_dims = [0],
            scatter_dims_to_operand_dims = [0], index_vector_dim = 2>,
        unique_indices = false}> ({
    ^bb0(%arg65: tensor<f32>, %arg66: tensor<f32>):
      %2093 = stablehlo.add %arg65, %arg66 : tensor<f32>
      stablehlo.return %2093 : tensor<f32>
    }) : (tensor<100352x2048xf32>, tensor<16x1024x1xi32>, tensor<16x1024x2048xf32>) -> tensor<100352x2048xf32>
    %2091 = sdy.sharding_constraint %2090 <@mesh, [{"model"}, {"data"}]> : tensor<100352x2048xf32>
    %2092 = stablehlo.add %3, %2091 : tensor<100352x2048xf32>
    ;                      ^^ dense grad    ^^ sparse grad
```

</details>

<details>
<summary><b>Optimized HLO: scatter fusions in the full step executable</b></summary>

After XLA optimization, the scatter-adds become `kCustom` fusions with `aliasing_operands` (in-place buffer reuse):

```
  %fusion.2 = f32[100352,256]{1,0:T(8,128)} fusion(
      %add_convert_fusion.1, %get-tuple-element.2480, %bitcast.5329, %broadcast_clamp_fusion.1),
      kind=kCustom, calls=%fused_computation.2433,
      metadata={op_name="jit(f_fun)/transpose(jvp(jit(forward)))/scatter-add" stack_frame_id=33},
      backend_config={..."aliasing_operands":{"lists":[{"indices":["0","4"]}]}}

  %fusion.3 = f32[100352,256]{1,0:T(8,128)} fusion(
      %add_convert_fusion, %get-tuple-element.2482, %bitcast.5330, %broadcast_clamp_fusion),
      kind=kCustom, calls=%fused_computation.2431,
      metadata={op_name="jit(f_fun)/transpose(jvp(jit(forward)))/scatter-add" stack_frame_id=84},
      backend_config={..."aliasing_operands":{"lists":[{"indices":["0","4"]}]}}
```

</details>

<details>
<summary><b>Optimized HLO: same scatter fusions in a grad-only executable (stable)</b></summary>

A grad-only executable (no optimizer update) has the same scatter fusions but merges them with a separate `add` instruction, and is always stable:

```
  %fusion.3 = f32[100352,256]{1,0:T(8,128)} fusion(...), kind=kCustom, ...
  %fusion.2 = f32[100352,256]{1,0:T(8,128)} fusion(...), kind=kCustom, ...
  %add_any.512 = f32[100352,256]{1,0:T(8,128)} add(%fusion.2, %fusion.3)
```

</details>

## System info (python version, jaxlib version, accelerator, etc.)

- TPU v6e-8
- `jax==0.9.0.1`
- `jaxlib==0.9.0.1`
- `optax==0.2.7`

I created a fresh TPU v6e-8 VM with the `v2-alpha-tpuv6e` image, and set up the environment as follows:
```bash
# install uv
curl -LsSf https://astral.sh/uv/install.sh | sh -s -- -q
source $HOME/.local/bin/env
uv venv ~/.venvs/py312 --python 3.12 -q
source ~/.venvs/py312/bin/activate

# install packages
uv pip install jax[tpu] optax
```

Output of `import jax; jax.print_environment_info()`:
```
jax:    0.9.0.1
jaxlib: 0.9.0.1
numpy:  2.4.2
python: 3.12.12 (main, Feb 12 2026, 00:42:14) [Clang 21.1.4 ]
device info: TPU v6 lite-8, 8 local devices"
process_count: 1
platform: uname_result(system='Linux', node='t1v-n-43e8b932-w-0', release='6.8.0-1015-gcp', version='#17~22.04.1-Ubuntu SMP Tue Sep  3 16:11:52 UTC 2024', machine='x86_64')
```

**Comment by rosspeili** [comment]
This is very detailed @martin-marek. Are you working on a solution, or could I get involved? Thanks.

**Comment by martin-marek** [comment]
@rosspeili I've been trying to narrow down the root cause and I updated the description. I'll see how far I can get but I feel like I'm a bit out of my depth here.

**Comment by martin-marek** [comment]
I submitted this issue and a proposed PR in https://github.com/openxla/xla/issues/38597
